File: scripts/legacy/fragment_attack.py
#!/usr/bin/env python3
"""
IDS Lab - Attaque par fragmentation IP avec Scapy
==================================================
Ce script démontre comment fragmenter des paquets pour éviter la détection IDS.

Usage:
    sudo python3 fragment_attack.py <target_ip> [fragment_size]

Exemple:
    sudo python3 fragment_attack.py 172.19.0.3 8
"""

import sys
import os

# Vérifier si on est root
if os.geteuid() != 0:
    print("⚠️  Ce script nécessite les privilèges root (pour envoyer des raw packets)")
    print("   Relancez avec: sudo python3 fragment_attack.py <target>")
    sys.exit(1)

try:
    from scapy.all import *
except ImportError:
    print("❌ Scapy n'est pas installé.")
    print("   Installez-le avec: pip3 install scapy")
    sys.exit(1)

def fragment_attack(target, frag_size=8):
    """
    Envoie une requête HTTP fragmentée en petits morceaux.

    Args:
        target: IP de la cible
        frag_size: Taille de chaque fragment (8 = minimum)
    """
    print(f"""
╔══════════════════════════════════════════════════════════════╗
║            Attaque par Fragmentation IP                      ║
╠══════════════════════════════════════════════════════════════╣
║  Target: {target}
║  Fragment size: {frag_size} bytes
╚══════════════════════════════════════════════════════════════╝
""")

    # Payload HTTP malveillant (simulé)
    payload = "GET /EVIL_PAYLOAD HTTP/1.1\r\nHost: target\r\nUser-Agent: NmapScan\r\n\r\n"

    # Créer le paquet complet
    ip = IP(dst=target)
    tcp = TCP(sport=RandShort(), dport=80, flags="S")

    # Premier test: SYN normal
    print("📍 Test 1: SYN normal (référence)")
    ans = sr1(ip/tcp, timeout=2, verbose=0)
    if ans:
        print(f"   Réponse: {ans.summary()}")
    else:
        print("   Pas de réponse (filtré ou fermé)")

    # Deuxième test: HTTP fragmenté
    print("\n📍 Test 2: Paquet HTTP fragmenté")

    # Construire le paquet complet
    full_pkt = ip/TCP(sport=RandShort(), dport=80, flags="PA")/Raw(load=payload)

    # Fragmenter manuellement
    fragments = fragment(full_pkt, fragsize=frag_size)

    print(f"   Nombre de fragments: {len(fragments)}")
    for i, frag in enumerate(fragments):
        print(f"   Fragment {i+1}: {len(frag)} bytes, offset={frag.frag}, MF={frag.flags.MF}")
        send(frag, verbose=0)

    print("\n✅ Fragments envoyés!")
    print("\n💡 Vérifiez les alertes dans EveBox: http://localhost:5636")
    print("   - Niveau 1-3: Probable que les fragments passent")
    print("   - Niveau 4-5: Fragmentation devrait être détectée")

def overlapping_fragments(target):
    """
    Technique avancée: fragments superposés pour confondre l'IDS.
    """
    print(f"\n📍 Test 3: Fragments superposés (overlapping)")

    ip = IP(dst=target)

    # Créer des fragments qui se chevauchent
    # Le premier fragment contient du padding innocent
    frag1 = ip/TCP(sport=12345, dport=80)/Raw(load="AAAA")
    frag1.flags = "MF"
    frag1.frag = 0

    # Le second fragment écrase une partie du premier avec le vrai payload
    frag2 = ip/TCP(sport=12345, dport=80)/Raw(load="GET /evil HTTP/1.1")
    frag2.frag = 1  # Overlap!

    send(frag1, verbose=0)
    send(frag2, verbose=0)

    print("   Fragments superposés envoyés!")
    print("   Cette technique exploite les différences de réassemblage IP/TCP")

def ttl_evasion(target):
    """
    Technique d'évasion par TTL: envoyer des paquets avec TTL court
    qui expirent avant l'IDS mais après le routeur.
    """
    print(f"\n📍 Test 4: Manipulation TTL")

    for ttl in [1, 3, 64, 128, 255]:
        pkt = IP(dst=target, ttl=ttl)/TCP(dport=80, flags="S")
        ans = sr1(pkt, timeout=1, verbose=0)
        status = "✓ Réponse" if ans else "✗ Timeout/TTL expiré"
        print(f"   TTL={ttl:3d}: {status}")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        # Essayer de trouver automatiquement la cible suricata
        try:
            import subprocess
            result = subprocess.run(
                ["docker", "inspect", "target_suricata", "--format",
                 "{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}"],
                capture_output=True, text=True
            )
            if result.returncode == 0 and result.stdout.strip():
                target = result.stdout.strip()
                print(f"🎯 Cible auto-détectée: {target}")
            else:
                print(__doc__)
                sys.exit(1)
        except:
            print(__doc__)
            sys.exit(1)
    else:
        target = sys.argv[1]

    frag_size = int(sys.argv[2]) if len(sys.argv) > 2 else 8

    fragment_attack(target, frag_size)
    overlapping_fragments(target)
    ttl_evasion(target)

File: scripts/lua/fragment_attack.lua
#!/usr/bin/env lua5.3
-- =============================================================================
-- IDS Lab - Attaque par fragmentation IP
-- =============================================================================
-- Ce script démontre des techniques d'évasion IDS par fragmentation.
--
-- NOTE: Pour des tests plus avancés (fragments superposés, manipulation TTL),
-- utilisez la version Python avec Scapy: scripts/legacy/fragment_attack.py
--
-- TECHNIQUES TESTÉES:
-- 1. Scan standard (référence)
-- 2. Fragmentation -f (8 bytes)
-- 3. Fragmentation --mtu 16
-- 4. Fragmentation --mtu 24
-- 5. Scan décoy (leurres)
-- 6. Timing lent (T1)
--
-- USAGE:
--   lua5.3 fragment_attack.lua [TARGET_IP] [OPTIONS]
--
-- OPTIONS:
--   --quick      Tests rapides uniquement
--   -h, --help   Afficher cette aide
-- =============================================================================

package.path = package.path .. ";./lib/?.lua"
local T = require("test_utils")

-- =============================================================================
-- CONFIGURATION
-- =============================================================================

local OPTIONS = {
    quick = false,
}

-- Chemin du projet
local SCRIPT_PATH = arg[0] or "."
local PROJECT_ROOT = T.get_project_root(SCRIPT_PATH)

-- =============================================================================
-- FONCTIONS
-- =============================================================================

local function get_suricata_ip()
    local cmd = "docker inspect target_suricata --format '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' 2>/dev/null"
    local output, exit_code = T.run(cmd)
    if exit_code == 0 and output ~= "" then
        return output:gsub("%s+", "")
    end
    return nil
end

local function run_scan(name, cmd)
    print("")
    print(T.colors.bold .. "📍 " .. name .. T.colors.reset)

    local output, exit_code = T.run(cmd .. " 2>/dev/null")

    if exit_code == 0 then
        for line in output:gmatch("[^\n]+") do
            if line:match("^PORT") or line:match("^%d+/") or line:match("open")
               or line:match("filtered") or line:match("closed") then
                print("   " .. line)
            end
        end
    else
        print("   " .. T.colors.yellow .. "Échec ou timeout" .. T.colors.reset)
    end
end

local function test_standard(target)
    run_scan(
        "Test 1: SYN Scan standard (référence)",
        string.format("sudo nmap -sS -p 80 --max-retries 1 -T4 %s", target)
    )
end

local function test_fragment_f(target)
    run_scan(
        "Test 2: Fragmentation -f (8 bytes)",
        string.format("sudo nmap -f -sS -p 80 --max-retries 1 -T4 %s", target)
    )
end

local function test_fragment_mtu16(target)
    run_scan(
        "Test 3: Fragmentation --mtu 16",
        string.format("sudo nmap --mtu 16 -sS -p 80 --max-retries 1 -T4 %s", target)
    )
end

local function test_fragment_mtu24(target)
    run_scan(
        "Test 4: Fragmentation --mtu 24",
        string.format("sudo nmap --mtu 24 -sS -p 80 --max-retries 1 -T4 %s", target)
    )
end

local function test_decoy(target)
    run_scan(
        "Test 5: Scan avec decoys (leurres)",
        string.format("sudo nmap -D RND:5 -sS -p 80 --max-retries 1 -T4 %s", target)
    )
end

local function test_slow(target)
    print("")
    print(T.colors.bold .. "📍 Test 6: Timing très lent (T1)" .. T.colors.reset)
    print("   (timeout 15s - T1 est très lent)")

    local cmd = string.format("timeout 15 sudo nmap -T1 -sS -p 80 --max-retries 0 %s 2>/dev/null", target)
    local output, _ = T.run(cmd)

    local found = false
    for line in output:gmatch("[^\n]+") do
        if line:match("open") or line:match("filtered") or line:match("closed") then
            print("   " .. line)
            found = true
        end
    end

    if not found then
        print("   " .. T.colors.yellow .. "Timeout (normal pour T1)" .. T.colors.reset)
    end
end

local function show_results()
    print("")
    print(T.colors.cyan .. "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━" .. T.colors.reset)
    print(T.colors.green .. "✅ Tests terminés!" .. T.colors.reset)
    print("")
    print(T.colors.bold .. "💡 Interprétation:" .. T.colors.reset)
    print("   - Niveau 1-3: La fragmentation devrait passer")
    print("   - Niveau 4-5: La fragmentation devrait être détectée")
    print("")
    print(T.colors.bold .. "📊 Vérifiez les alertes:" .. T.colors.reset)
    print("   - EveBox: http://localhost:5636")
    print("   - Kibana: http://localhost:5601")
    print("")
    print(T.colors.bold .. "🔬 Pour des tests avancés:" .. T.colors.reset)
    print("   - Fragments superposés (overlapping)")
    print("   - Manipulation TTL")
    print("   Utilisez: sudo python3 scripts/legacy/fragment_attack.py")
    print("")
end

-- =============================================================================
-- MAIN
-- =============================================================================

local function main()
    local target = nil

    -- Parse arguments
    for i = 1, #arg do
        if arg[i] == "-h" or arg[i] == "--help" then
            print([[
Usage: lua5.3 fragment_attack.lua [TARGET_IP] [OPTIONS]

Arguments:
  TARGET_IP    IP de la cible (auto-détecté si non spécifié)

Options:
  --quick      Tests rapides uniquement (skip T1)
  -h, --help   Afficher cette aide

Techniques testées:
  1. Scan standard (référence)
  2. Fragmentation -f (8 bytes)
  3. Fragmentation --mtu 16
  4. Fragmentation --mtu 24
  5. Scan décoy (leurres)
  6. Timing lent (T1)

Exemple:
  sudo lua5.3 fragment_attack.lua 172.29.0.100
  sudo lua5.3 fragment_attack.lua --quick

Note: Pour des tests plus avancés (fragments superposés, TTL),
utilisez: sudo python3 scripts/legacy/fragment_attack.py
]])
            os.exit(0)
        elseif arg[i] == "--quick" then
            OPTIONS.quick = true
        elseif not arg[i]:match("^%-") then
            target = arg[i]
        end
    end

    -- Auto-détecter la cible
    if not target then
        target = get_suricata_ip()
        if target then
            print(T.colors.cyan .. "🎯 Cible auto-détectée: " .. target .. T.colors.reset)
        end
    end

    if not target then
        print([[
Usage: lua5.3 fragment_attack.lua [TARGET_IP]

Auto-détection échouée. Spécifiez l'IP de la cible:
  lua5.3 fragment_attack.lua 172.29.0.100

Ou démarrez le lab Suricata:
  cd suricata-lab && docker compose up -d
]])
        os.exit(1)
    end

    -- Header
    print(T.colors.cyan .. T.colors.bold)
    print("╔══════════════════════════════════════════════════════════════╗")
    print("║            Attaque par Fragmentation IP                      ║")
    print("╠══════════════════════════════════════════════════════════════╣")
    print(string.format("║  Target: %-51s ║", target))
    print("╚══════════════════════════════════════════════════════════════╝")
    print(T.colors.reset)

    -- Vérifier les privilèges
    local uid_output, _ = T.run("id -u")
    local uid = tonumber(uid_output:gsub("%s+", "")) or 1000

    if uid ~= 0 then
        T.log_warning("Ce script nécessite les privilèges root pour les scans SYN")
        print("   Relancez avec: sudo lua5.3 fragment_attack.lua " .. target)
        print("")
    end

    -- Tests
    test_standard(target)
    test_fragment_f(target)
    test_fragment_mtu16(target)
    test_fragment_mtu24(target)
    test_decoy(target)

    if not OPTIONS.quick then
        test_slow(target)
    else
        print("")
        print(T.colors.yellow .. "   Test T1 ignoré (--quick)" .. T.colors.reset)
    end

    show_results()
end

main()
